Skip executing fetch handlers if it is no-op.

yoshisatoyanagisawa opened the pull request:
This is the specification change for the proposal (https://github.com/w3c/ServiceWorker/issues/1671).

I got comments and/or contribution from following folks when I was drafting this specification update:
+@jakearchibald 
+@hiroshige-g

Let me ask feedbacks from the WG.
Thank you.


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/yoshisatoyanagisawa/ServiceWorker/pull/1672.html" title="Last updated on Mar 17, 2023, 9:11 AM UTC (0d63641)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/ServiceWorker/1672/25f60ac...yoshisatoyanagisawa:0d63641.html" title="Last updated on Mar 17, 2023, 9:11 AM UTC (0d63641)">Diff</a>

jakearchibald commented on docs/index.bs:
```diff
@@ -2970,16 +2972,34 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+                  Note: If the global object's associated list of event listeners does not have any event listener added at this moment, the service worker's set of event types to handle remains an empty set.
 
               1. Set |script|'s <a>has ever been evaluated flag</a>.
+              1. If the [=No-Op Handler identification=] algorithm with |serviceWorker| returns true, the [=has had non-no-op fetch listeners flag=] is set to false.
```
This seems backwards. I don't think we ever want to unset this flag. Or have I misunderstood the design?


```suggestion
              1. If [=No-Op Handler identification=] given |serviceWorker| is false, then set |serviceWorker|'s [=has had non-no-op fetch listeners flag=].
```

When referring to a property like `[=has had non-no-op fetch listeners flag=]`, be clear about which object you're referring to.

yoshisatoyanagisawa replied:
Let me update the flag name.  The current flag name requires to have the default value because we should not think fetch handlers are no-op by default.

jakearchibald commented on docs/index.bs:
```diff
@@ -2970,16 +2972,34 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+      : Output
+      :: a boolean
+
+      1. If |service worker|'s set of event types to handle does not contain "fetch", then returns false.
```
`|service worker|` should be `|serviceWorker|` (as defined in the input)

References need to be linked properly here.

yoshisatoyanagisawa replied:
done.

jakearchibald commented on docs/index.bs:
```diff
@@ -2970,16 +2972,34 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+      :: a boolean
+
+      1. If |service worker|'s set of event types to handle does not contain "fetch", then returns false.
+      1. If the callback's function bodies (https://tc39.es/ecma262/#prod-FunctionBody) of all fetch event listener in |service worker| are empty, then the user agent *may* return true.
```
Things like "callback" need to be referenced properly. Eg, see https://dom.spec.whatwg.org/#event-listener-callback - you'll need to handle cases where the 'callback' is actually an `EventListener` object.

`|service worker|` should be `|serviceWorker|` (as defined in the input)

Rather than just paste the link of https://tc39.es/ecma262/#prod-FunctionBody, create a proper reference. See the `<pre class="anchors">` section at the top of the document.

"of all fetch event listener" needs to be defined correctly in terms of https://dom.spec.whatwg.org/#concept-event-listener.

"empty" needs to be defined more robustly.

yoshisatoyanagisawa replied:
I am not so confident to write references, but I tried to address this.  please take a look.

yoshisatoyanagisawa replied:
"event listeners" seems to be already defined.  Do you want to replace all of them with DOM standard's event listener?

yoshisatoyanagisawa replied:
To avoid the conflict of the "event listeners" name, I avoided to use it while referring callback.

yoshisatoyanagisawa replied:
Please feel free to reopen if this still matters.

jakearchibald commented on docs/index.bs:
```diff
@@ -2970,16 +2972,34 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+      1. If |service worker|'s set of event types to handle does not contain "fetch", then returns false.
+      1. If the callback's function bodies (https://tc39.es/ecma262/#prod-FunctionBody) of all fetch event listener in |service worker| are empty, then the user agent *may* return true.
+
+      Note: it detects something like `onfetch = () => {}`. Some sites have a fetch event listener with empty body to make them recognized as progressive web application (PWA).
```
This comment should be expanded upon, but I can help with that.

jakearchibald commented on docs/index.bs:
```diff
@@ -2970,16 +2972,34 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+
+      Note: it detects something like `onfetch = () => {}`. Some sites have a fetch event listener with empty body to make them recognized as progressive web application (PWA).
+
+      1. return false.
```
```suggestion
      1. Return false.
```

yoshisatoyanagisawa replied:
done.

jakearchibald commented on docs/index.bs:
```diff
@@ -3041,7 +3061,7 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
           1. If |registration| is null or |registration|'s <a>active worker</a> is null, return null.
           1. If |request|'s [=request/destination=] is not {{RequestDestination/"report"}}, set |reservedClient|'s <a>active service worker</a> to |registration|'s <a>active worker</a>.
-          1. If |request| is a [=navigation request=], |registration|'s [=navigation preload enabled flag=] is set, |request|'s [=request/method=] is \`<code>GET</code>\`, and |registration|'s [=active worker=]'s <a>set of event types to handle</a> [=set/contains=] <code>fetch</code>, then:
+          1. If |request| is a [=navigation request=], |registration|'s [=navigation preload enabled flag=] is set, |request|'s [=request/method=] is \`<code>GET</code>\`, |registration|'s [=active worker=]'s <a>set of event types to handle</a> [=set/contains=] <code>fetch</code>, and the result of running [=Skippable Fetch Handler=] algorithm with [=active worker=] is false then:
```
```suggestion
          1. If |request| is a [=navigation request=], |registration|'s [=navigation preload enabled flag=] is set, |request|'s [=request/method=] is \`<code>GET</code>\`, |registration|'s [=active worker=]'s <a>set of event types to handle</a> [=set/contains=] <code>fetch</code>, and the result of running [=Skippable Fetch Handler=] algorithm with |registration|'s [=active worker=] is false then:
```

yoshisatoyanagisawa replied:
done.

jakearchibald commented on docs/index.bs:
```diff
@@ -3077,6 +3097,12 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
       1. If the result of running the [=Should Skip Event=] algorithm with "fetch" and |activeWorker| is true, then:
           1. If |shouldSoftUpdate| is true, then [=in parallel=] run the [=Soft Update=] algorithm with |registration|.
           1. Return null.
+      1. If the result of running [=Skippable Fetch Handler=] algorithm with |activeWorker| is true, then:
```
```suggestion
      1. If the result of running the [=Skippable Fetch Handler=] algorithm with |activeWorker| is true, then:
```

yoshisatoyanagisawa replied:
done.

jakearchibald commented on docs/index.bs:
```diff
@@ -3077,6 +3097,12 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
           1. If |shouldSoftUpdate| is true, then [=in parallel=] run the [=Soft Update=] algorithm with |registration|.
           1. Return null.
+      1. If the result of running [=Skippable Fetch Handler=] algorithm with |activeWorker| is true, then:
+          1. run the following steps [=in parallel=]:
```
```suggestion
          1. [=In parallel=]:
```

yoshisatoyanagisawa replied:
done.

jakearchibald commented on docs/index.bs:
```diff
@@ -3077,6 +3097,12 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
           1. Return null.
+      1. If the result of running [=Skippable Fetch Handler=] algorithm with |activeWorker| is true, then:
+          1. run the following steps [=in parallel=]:
+              1. If |activeWorker|'s state is "activating", wait for |activeWorker|'s state to become "activated".
```
```suggestion
              1. If |activeWorker|'s [=service worker/state=] is "activating", then wait for |activeWorker|'s [=service worker/state=] to become "activated".
```

Make sure terms you use are referenced

yoshisatoyanagisawa replied:
done.

jakearchibald commented on docs/index.bs:
```diff
@@ -191,6 +191,8 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
 
     A [=/service worker=] has an associated <dfn>start status</dfn> which can be null or a [=Completion=]. It is initially null.
 
+    A [=/service worker=] has an associated <dfn>has had non-no-op fetch listeners flag</dfn>. It is initially true.
```
Maybe rename this to reflect the uncertain nature?

```suggestion
    A [=/service worker=] has an associated <dfn>may have had non-empty fetch listener flag</dfn>. It is initially unset.
```

yoshisatoyanagisawa replied:
I have renamed the flag without negation.  Please let me know what you think.

yoshisatoyanagisawa replied:
Please feel free to reopen if this still matters.

jakearchibald commented on docs/index.bs:
```diff
@@ -3132,6 +3158,17 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+      : Output
+      :: a boolean
+
+      1. If |serviceworker| has |has non-no-op fetch listeners flag|, and the value is false, then the user agent *may* return true.
```
What's the reason for "may" here, given that "No-Op Handler identification" already uses "may"?

yoshisatoyanagisawa replied:
I just tried to follow what "Should Skip Event" does, but it does not need to be *may*.  Let me omit.

annevk commented:
I suspect that event handlers might be tricky here as while an event handler can be considered empty, the event listener it ends up adding won't be (as it wraps the function). Perhaps this can be addressed by poking at HTML's event handler map. Gets a little hairy though.

Might be good to get some advice from @domenic and possibly some JS folks.

jakearchibald commented:
@yoshisatoyanagisawa it's been a while since I worked on the service worker spec, and it's built up some errors over time as other specs have changed, and bikeshed has updated. I'm going to fix up all those errors to make it easier for you to make this change.

jakearchibald commented:
@ yoshisatoyanagisawa I've got the errors down to 14 (I need changes in other specs to land before I can make it 0). Hopefully the conflicts aren't too much trouble.

I'm going to take a closer look at this PR tomorrow.

yoshisatoyanagisawa commented:
@jakearchibald thank you.  I have resolved a merge conflict, and fixed bikeshed errors caused within my change.  Please take another look.

domenic commented on docs/index.bs:
```diff
@@ -2973,16 +2983,44 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+      1. set |eventListenerList| to [=map/value=] whose [=map/key=] is <code>fetch</code> in |workerGlobalScope|'s <a>set of event types to handle</a>.
+      1. [=list/For each=] |eventListener| of |eventListenerList|:
+          1. If |eventListener|'s [=callback=] is not null:
+              1. Set |callback| to an ECMAScript object coverted from |eventListener|'s [=callback=], which is [=idl-callback-interface=].
```
```suggestion
              1. Set |callback| to the result of [=converting|convert to an ECMAScript value=] |eventListener|'s [=callback=] to an ECMAScript value.
```

yoshisatoyanagisawa replied:
done.

domenic commented on docs/index.bs:
```diff
@@ -2973,16 +2983,44 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+      1. [=list/For each=] |eventListener| of |eventListenerList|:
+          1. If |eventListener|'s [=callback=] is not null:
+              1. Set |callback| to an ECMAScript object coverted from |eventListener|'s [=callback=], which is [=idl-callback-interface=].
+              1. If |callback| is a [=function=], and [=function body=] is not empty (i.e. either [=statement=] or [=declaration=] exist), then return false.
```
```suggestion
              1. If [$IsCallable$](|callback|), and |callback|'s [=function body=] is not empty (i.e. either a [=statement=] or [=declaration=] exist), then return false.
```

yoshisatoyanagisawa replied:
I followed the suggestion, and converted the statement to two statements.

domenic commented on docs/index.bs:
```diff
@@ -2973,16 +2983,44 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+          1. If |eventListener|'s [=callback=] is not null:
+              1. Set |callback| to an ECMAScript object coverted from |eventListener|'s [=callback=], which is [=idl-callback-interface=].
+              1. If |callback| is a [=function=], and [=function body=] is not empty (i.e. either [=statement=] or [=declaration=] exist), then return false.
+              1. If |callback| does not have <code>handleEvent</code> property, then return false.
+              1. If |callback|'s <code>handleEvent</code> property is not a [=function=], then return false.
+              1. If a [=function body=] of |callback|'s <code>handleEvent</code> property is not empty (i.e. either [=statement=] or [=declaration=] exist), return false.
```
```suggestion
              1. Return false.
```

yoshisatoyanagisawa replied:
Considering https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/bindings/core/v8/js_event_listener.cc;l=14;drc=609b81c1b05359d844df8786169ba20f230152b4 behavior, it might be fair to say that this would also see handleEvent.  I tried to update this with  https://webidl.spec.whatwg.org/#call-a-user-objects-operation 10 upon your offline suggestion.

domenic reviewed (commented):
Some suggested tweaks to be more precise, plus a suggestion that you bail out in the object case instead of potentially triggering side effects from the `handleEvent` getter.

domenic commented on docs/index.bs:
```diff
@@ -62,6 +64,8 @@ spec: ecma-262; urlPrefix: http://tc39.github.io/ecma262/
         text: NormalCompletion; url: sec-normalcompletion
         text: ThrowCompletion; url: sec-throwcompletion
+        text: IsCallable; url: sec-iscallable
+        text: Get; url: sec-get-o-p
```
These should not be necessary these days. But, maybe they are; if removing them breaks then ignore my advice :)

yoshisatoyanagisawa replied:
They seem to be needed. 
LINE ~3011: No 'abstract-op' refs found for 'IsCallable'.
[$IsCallable$]
LINE ~3012: No 'abstract-op' refs found for 'Get'.
[$Get$]
LINE ~3015: No 'abstract-op' refs found for 'IsCallable'.
[$IsCallable$]

domenic commented on docs/index.bs:
```diff
@@ -2973,16 +2983,46 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+      :: a boolean
+
+      1. If |workerGlobalScope|'s <a>set of event types to handle</a> does not [=set/contain=] <code>fetch</code>, then returns false.
+      1. set |eventListenerList| to [=map/value=] whose [=map/key=] is <code>fetch</code> in |workerGlobalScope|'s <a>set of event types to handle</a>.
```
The inner loop below this line now looks great.

This step doesn't seem correct though. |eventListenerList| should be https://dom.spec.whatwg.org/#eventtarget-event-listener-list . And then you need to filter out the ones where https://dom.spec.whatwg.org/#event-listener-type is not "fetch".

yoshisatoyanagisawa replied:
I misunderstood the mental model of "set of event types to handle".  I thought is is something like std::map, but it is actually set.  I have updated the explanation like so.

jakearchibald commented on docs/index.bs:
```diff
@@ -2973,16 +2984,48 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+
+      1. If |workerGlobalScope|'s <a>set of event types to handle</a> does not [=set/contain=] <code>fetch</code>, then returns false.
+      1. Let |eventListenerList| be an empty [=list=].
+      1. [=list/For each=] |eventListener| of |workerGlobalScope|'s <a>set of event types to handle</a>:
```
This isn't right. `<a>set of event types to handle</a>` is a set of strings, not listeners.

In cases like this, where I want to get something like "all event listener callbacks of a particular type", I try to think of existing things that do that already that I could take inspiration from. In this case, I'd look at `target.dispatchEvent`, since it has to run all of the callbacks of a particular type.

So, I'd start at https://dom.spec.whatwg.org/#dom-eventtarget-dispatchevent, and follow the algorithm, and see how they gather up the callbacks. We probably need to do a similar thing here.

Following the trail takes me to https://dom.spec.whatwg.org/#concept-event-listener-invoke, which shows me how to get the listeners for a particular event target.

Then it goes to https://dom.spec.whatwg.org/#concept-event-listener-inner-invoke, which shows me how to filter by event type.

Then it goes to https://webidl.spec.whatwg.org/#call-a-user-objects-operation which shows me how to get the callback from the two different forms of listener.

yoshisatoyanagisawa replied:
Hmm, I got stuck on finding who sets currentTarget (https://dom.spec.whatwg.org/#dom-event-currenttarget) in Dispatching Events (https://dom.spec.whatwg.org/#concept-event-dispatch).
If that is cleared, we can get event listeners from currentTarget like invoke's step 6 (https://dom.spec.whatwg.org/#concept-event-listener-invoke), and the current algorithm can be used.
I tired to update anyway.

jakearchibald commented on docs/index.bs:
```diff
@@ -3130,6 +3179,17 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+      : Output
+      :: a boolean
+
+      1. If |serviceWorker| has [=all fetch listeners are empty flag=] and set, then the user agent *may* return true.
```
I think the "may" condition should sit in "Empty Handler identification", since that's the part of the process that's optional.

yoshisatoyanagisawa replied:
Sure.  I have moved the part to "Empty Fetch Handler Identification" algorithm.

yoshisatoyanagisawa commented on docs/index.bs:
```diff
@@ -2973,16 +2984,48 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
                   1. [=set/Append=] |eventType| to |workerGlobalScope|'s associated [=ServiceWorkerGlobalScope/service worker=]'s <a>set of event types to handle</a>.
 
-                  Note: If the global object's associated list of event listeners does not have any event listener added at this moment, the service worker's set of event types to handle remains an empty set. The user agents are encouraged to show a warning that the event listeners must be added on the very first evaluation of the worker script.
+                  Note: If the global object's associated list of event listeners does not have any event listener added at this moment, the service worker's <a>set of event types to handle</a> remains an empty set.
```
done.

yoshisatoyanagisawa commented on docs/index.bs:
```diff
@@ -2973,16 +2984,48 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
   </section>
 
+  <section algorithm>
+    <h3 id="empty-handler-identification-algorithm"><dfn>Empty Handler identification</dfn></h3>
```
added "fetch".

yoshisatoyanagisawa commented on docs/index.bs:
```diff
@@ -2973,16 +2984,48 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+                  Note: If the global object's associated list of event listeners does not have any event listener added at this moment, the service worker's <a>set of event types to handle</a> remains an empty set.
 
               1. Set |script|'s <a>has ever been evaluated flag</a>.
+              1. If the [=Empty Handler identification=] algorithm with |workerGlobalScope| returns true, the [=all fetch listeners are empty flag=] is set.
```
I think you are talking about [Approaches to deal with the handler updates after the initialization](https://github.com/yoshisatoyanagisawa/service-worker-skip-no-op-fetch-handler#approaches-to-deal-with-the-handler-updates-after-the-initialization) in the explainer.  This implements [Option 2](https://github.com/yoshisatoyanagisawa/service-worker-skip-no-op-fetch-handler#option-2-updates-are-checked-every-time-when-the-service-worker-starts).  However, this just set "all fetch listeners are empty flag", and never unset.  Upon the algorithm, once the flag is set, it would be kept and does not behave like Option 2 (i.e. the fetch handler function body is evaluated every time service worker starts).  Let me fix.

yoshisatoyanagisawa commented on docs/index.bs:
```diff
@@ -2973,16 +2984,48 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+      : Output
+      :: a boolean
+
+      1. If |workerGlobalScope|'s <a>set of event types to handle</a> does not [=set/contain=] <code>fetch</code>, then returns false.
```
The behavior is intended, but people may feel the flag name does not match with the behavior.
I intend to say the fetch handler exists and all of them are empty, but it sounds long.

annevk commented on docs/index.bs:
```diff
@@ -2973,16 +2984,49 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
   </section>
 
+  <section algorithm>
+    <h3 id="empty-fetch-handler-identification-algorithm"><dfn>Empty Fetch Handler Identification</dfn></h3>
+
+      : Input
+      :: |workerGlobalScope|, a [=service worker/global object=].
+      : Output
+      :: a boolean
+
+      1. If |workerGlobalScope|'s [=set of event types to handle=] does not [=set/contain=] <code>fetch</code>, then returns false.
+      1. Let |eventListenerList| be an empty [=list=].
+      1. [=list/For each=] |eventListener| of |workerGlobalScope|'s [=set of event types to handle=]:
+          1. If |eventListener|'s <a spec="dom">type</a> is <code>fetch</code>, then [=list/append=] |eventListener| to |eventListenerList|.
+      1. [=list/For each=] |eventListener| of |eventListenerList|:
+          1. If |eventListener|'s [=callback=] is not null:
+              1. Set |callback| to the result of [=converting|convert to an ECMAScript value=] |eventListener|'s [=callback=] to an ECMAScript value.
+              1. If [$IsCallable$](|callback|) is false:
+                  1. Let |getResult| be [=Completion=]([$Get$](|callback|), <code>handleEvent</code>).
+                  1. If |getResult| is [=abrupt completion=], then return false.
+                  1. Set |callback| to |getResult|.
+              1. If [$IsCallable$](|callback|) is false, then return false.
+              1. If |callback|'s [=function body=] is not empty (i.e. either a [=statement=] or [=declaration=] exist), then return false.
+
+              Note: it detects something like `onfetch = () => {}`. Some sites have a fetch event listener with empty body to make them recognized as progressive web application (PWA).
```
@domenic I think this still doesn't work for `onfetch` as HTML wraps that function in its own function when creating the callback, right? See https://github.com/WebKit/standards-positions/issues/129#issuecomment-1431008116.

domenic replied:
Yes. We started discussing how to fix that (offline) but then got sidetracked into making the surrounding parts of the spec more precise.

To detech `onfetch = () => {}` I think the following would work, nested immediately under "If _eventListener_'s callback is not null":

> 1. If _eventListener_ is an `EventListener` of the type created by [activate an event handler](https://html.spec.whatwg.org/#activate-an-event-handler), then set |callback| to the result of [converting] _eventTarget_'s [event handler map](https://html.spec.whatwg.org/#event-handler-map)["`fetch`"]'s [value](https://html.spec.whatwg.org/#event-handler-value) to an ECMAScript value.
> 2. Otherwise, set |callback| to ...

(you could also factor out the [converting] into a separate step.)

yoshisatoyanagisawa replied:
Can I ask where eventTarget in Step 1 come from?

yoshisatoyanagisawa replied:
I am not fully understand the comment, but I tried to update the steps as suggested.
https://github.com/w3c/ServiceWorker/pull/1672/commits/27f0e006edf2af02c1c11a9421879aa11f467c51

yoshisatoyanagisawa commented:
Can I ask some more comments on how the empty fetch handler identification algorithm should be?  I am also happy to listen any other comments if there is.

jakearchibald commented on docs/index.bs:
```diff
@@ -2973,16 +2984,52 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+          1. If |eventListener|'s <a spec="dom">type</a> is <code>fetch</code>, then [=list/append=] |eventListener| to |eventListenerList|.
+      1. [=list/For each=] |eventListener| of |eventListenerList|:
+          1. If |eventListener|'s [=callback=] is not null:
+              1. If |eventListener| is an EventListener of the type created by <a spec="html">activate an event handler</a>, then set |callback| to the result of [=converting|convert to an ECMAScript value=] |workerGlobalScope|'s <a spec="html">event handler map</a>["fetch"]'s value to an ECMAScript value.
```
@domenic ignoring the formatting issues, are you ok with this kind of hand wave? Alternative language could be: "Get the [event handler](https://html.spec.whatwg.org/multipage/webappapis.html#event-handlers) with a [listener](https://html.spec.whatwg.org/multipage/webappapis.html#event-handler-listener) of |eventListener|'s [=callback=], or null if none exists."

Then we can pull the value from that object.

domenic replied:
I suggested this type of handwave, because I didn't realize we could do better. But I think you're right. I think it should be something like this:

- Outside the loop, let |eventHandler| be |workerGlobalScope|'s <a spec="html">event handler map</a>["fetch"].

- Inside the loop, check if |eventListener| equals |eventHandler|'s [listener](https://html.spec.whatwg.org/multipage/webappapis.html#event-handler-listener). If so, set |callback| to the result of converting |eventHandler|'s [value](https://html.spec.whatwg.org/multipage/webappapis.html#event-handler-value) to an ECMAScript value.

jakearchibald commented on docs/index.bs:
```diff
@@ -3130,6 +3183,17 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
       1. Return |response|.
   </section>
 
+  <section algorithm>
+    <h3 id="skippable-fetch-handler-algorithm"><dfn>Skippable Fetch Handler</dfn></h3>
+      : Input
+      :: |serviceWorker|, a [=/service worker=]
+      : Output
+      :: a boolean
+
+      1. If |serviceWorker| has [=all fetch listeners are empty flag=] and set, then the user agent return true.
+      1. Return false.
+  </section>
```
I still don't think we need this. It's just returning an existing boolean value.

domenic commented on docs/index.bs:
```diff
@@ -2973,16 +2984,52 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
 
               1. Set |script|'s <a>has ever been evaluated flag</a>.
+              1. Unset the [=all fetch listeners are empty flag=].
+              1. If the [=Empty Fetch Handler Identification=] algorithm with |workerGlobalScope| returns true, the [=all fetch listeners are empty flag=] may be set.
```
I like your suggested wording!

jakearchibald commented on docs/index.bs:
```diff
@@ -2973,16 +2984,52 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
   </section>
 
+  <section algorithm>
+    <h3 id="all-fetch-handlers-are-empty-algorithm"><dfn>All Fetch Handlers Are Empty</dfn></h3>
```
Ugh, my bad, we should be using "listeners" rather than "handlers". I should have caught this in the last review, sorry.

jakearchibald commented on docs/index.bs:
```diff
@@ -2973,16 +2985,51 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+      : Output
+      :: a boolean
+
+      1. If |workerGlobalScope|'s [=set of event types to handle=] does not [=set/contain=] <code>fetch</code>, then returns true.
```
```suggestion
      1. If |workerGlobalScope|'s [=set of event types to handle=] does not [=set/contain=] <code>fetch</code>, then return true.
```

jakearchibald commented on docs/index.bs:
```diff
@@ -2973,16 +2985,51 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+      :: a boolean
+
+      1. If |workerGlobalScope|'s [=set of event types to handle=] does not [=set/contain=] <code>fetch</code>, then returns true.
+      1. Let |eventHandler| be |workerGlobalScope|'s <a spec="html">event handler map</a>["fetch"]'s value.
```
```suggestion
      1. Let |eventHandler| be |workerGlobalScope|'s <a spec="html">event handler map</a>["onfetch"]'s value.
```

As per https://html.spec.whatwg.org/multipage/webappapis.html#event-handler-name, handler names always start "`on`".

jakearchibald commented on docs/index.bs:
```diff
@@ -2973,16 +2985,51 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+      1. Let |eventHandler| be |workerGlobalScope|'s <a spec="html">event handler map</a>["fetch"]'s value.
+      1. Let |eventListenerList| be an empty [=list=].
+      1. [=list/For each=] |eventListener| of |workerGlobalScope|'s [=event listener list=]:
+          1. If |eventListener|'s <a spec="dom">type</a> is <code>fetch</code>, then [=list/append=] |eventListener| to |eventListenerList|.
```
```suggestion
          1. If |eventListener|'s <a spec="dom">type</a> is "<code>fetch</code>", then [=list/append=] |eventListener| to |eventListenerList|.
```

jakearchibald commented on docs/index.bs:
```diff
@@ -2973,16 +2985,51 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+      1. [=list/For each=] |eventListener| of |workerGlobalScope|'s [=event listener list=]:
+          1. If |eventListener|'s <a spec="dom">type</a> is <code>fetch</code>, then [=list/append=] |eventListener| to |eventListenerList|.
+      1. [=list/For each=] |eventListener| of |eventListenerList|:
+          1. If |eventListener|'s [=callback=] is not null:
```
```suggestion
          1. If |eventListener|'s [=event listener/callback=] is null, then [=continue=].
```

Then un-indent the subsequent lines. This is mostly personal preference. I prefer early-return/continue rather than nesting everything in an `if`.

I swapped `[=callback=]` for `[=event listener/callback=]`. It isn't an issue now, but it feels like this could become ambiguous in future.

jakearchibald commented on docs/index.bs:
```diff
@@ -2973,16 +2985,51 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+          1. If |eventListener|'s <a spec="dom">type</a> is <code>fetch</code>, then [=list/append=] |eventListener| to |eventListenerList|.
+      1. [=list/For each=] |eventListener| of |eventListenerList|:
+          1. If |eventListener|'s [=callback=] is not null:
+              1. If |eventListener| equals |eventHandler|'s <a spec="html">listener</a>, then set |callback| to the result of [=converting|convert to an ECMAScript value=] |eventHandler|'s value to an ECMAScript value.
```
"|eventHandler|" may be null here.

Also, "value" in "|eventHandler|'s value" needs to be referenced properly.

jakearchibald commented on docs/index.bs:
```diff
@@ -2973,16 +2985,51 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+              1. If [$IsCallable$](|callback|) is false, then return false.
+              1. If |callback|'s [=function body=] is not empty (i.e. either a [=statement=] or [=declaration=] exist), then return false.
+
+              Note: it detects something like `onfetch = () => {}`. Some sites have a fetch event listener with empty body to make them recognized as progressive web application (PWA).
```
```suggestion
              Note: This detects "<code>fetch</code>" listeners like `() => {}`. Some sites have a fetch event listener with empty body to make them recognized by Chromium as a progressive web application (PWA).
```

jakearchibald commented on docs/index.bs:
```diff
@@ -2973,16 +2985,51 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+
+      1. Return true.
+
+      Note: the user agents are encouraged to show warnings that the event listeners are no-op, and their executions can be skipped.
```
```suggestion
      Note: User agents are encouraged to show a warning indicating that empty "<code>fetch</code>" listeners are unnecessary, and may have a negative performance impact.
```

jakearchibald reviewed (changes requested):
This is really close to ready!

The PR uses `<a spec="">…</a>` links to link to concepts that haven't been exported by other specs. Instead, PRs should be made against those specs to export those terms. Example: https://github.com/whatwg/html/pull/8924

yoshisatoyanagisawa commented:
I hope to touch other specs for exporting terms after this PR.

yoshisatoyanagisawa commented:
I believe I have addressed all existing comments.  Are there anything to change?

jakearchibald commented on docs/index.bs:
```diff
@@ -2973,16 +2985,51 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+          1. If |eventListener|'s <a spec="dom">type</a> is "<code>fetch</code>", then [=list/append=] |eventListener| to |eventListenerList|.
+      1. [=list/For each=] |eventListener| of |eventListenerList|:
+          1. If |eventListener|'s [=event listener/callback=] is null, then [=continue=].
+          1. If |eventHandler| is not null, and |eventListener| equals |eventHandler|'s <a spec="html">listener</a>, then set |callback| to the result of [=converting|convert to an ECMAScript value=] |eventHandler|'s <a spec="html">listener</a> value to an ECMAScript value.
```
Lines like this are still pointing to unexported terms in other specs. Can you get these exported in other specs first? I don't want to land the spec linking to unexported terms.

jakearchibald commented on docs/index.bs:
```diff
@@ -3075,6 +3122,12 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
+          1. [=In parallel=]:
+              1. If |activeWorker|'s [=service worker/state=] is "activating", then wait for |activeWorker|'s [=service worker/state=] to become "activated".
+              1. Run the [=Run Service Worker=] algorithm with |activeWorker|.
+              1. If |shouldSoftUpdate| is true, then run the [=Soft Update algorithm=] with |registration|.
```
```suggestion
              1. If |shouldSoftUpdate| is true, then run the [=Soft Update=] algorithm with |registration|.
```

The link text here was wrong (and was causing a bikeshed error.

jakearchibald reviewed (changes requested):
One small error, and the exporting from other specs left to do.

yoshisatoyanagisawa commented:
Will you take another look at the PR?  I believe I have addressed your comments.
I suppose an error like "LINE ~3010: No 'dfn' refs found for 'event handler map' that are marked for export." is because the time lag between the PR is merged and available, but I am not sure.  I hope I can also have advice on that.

Thank you.

jakearchibald commented:
Sighhh, something git-related went wrong here. Let me repair…

jakearchibald commented:
Continuing in https://github.com/w3c/ServiceWorker/pull/1674
